package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,262 @@
+            This selection of atoms is used to calculate the RMSD between
+            different frames. Water should be excluded.
+        epsilon : float, optional
+            Specifies the epsilon used for the diffusion map. More information
+            in [1] and [2]. With 'average' the average of the RMSD to the k-th
+            greatest value is used. (#TODO, check if actually agrees with theory)
+        k : int, optional
+            specifies the k for the epsilon choice if average epsilon is
+            used.
+        weights: list, optional
+            The list has to have the same length as the trajectory.
+            With 'None' the weight of each frame of the trajectory will be the
+            same.
+        metric : function, optional
+            Maps two numpy arrays to a float, is positive definite and symmetric,
+            Default: ``None`` sets metric to qcp.CalcRMSDRotationalMatrix
+        start : int, optional
+            First frame of trajectory to analyse, Default: 0
+        stop : int, optional
+            Last frame of trajectory to analyse, Default: -1
+        step : int, optional
+            Step between frames to analyse, Default: 1
+        """
+        self._u = u
+        self.atoms = u.select_atoms(select)
+        self._natoms = self.atoms.n_atoms
+        # modulus to prevent index out of bounds exception
+        traj = u.trajectory
+        self._epsilon = epsilon
+        if metric is not None:
+            self._metric = metric
+        else:
+            self._metric = qcp.CalcRMSDRotationalMatrix
+        # remember that this must be called before referencing self.nframes
+        self._setup_frames(traj, start, stop, step)
+
+        # modulus to prevent array index out of bounds exception
+        self._k = k % self.nframes
+        if weights is None:
+            # weights do not apply to metric
+            self._weights = np.ones((self.nframes,))
+        else:
+            if weights.shape[0] != self.nframes:
+                raise ValueError("The weight should have the same length as the"
+                                 'trajectory')
+            else:
+                # weights are constructed as relative to the mean
+                self._weights = (np.asarray(weights, dtype=np.float64) /
+                                 np.mean(weights))
+
+    def _prepare(self):
+        self.diffusion_matrix = np.zeros((self.nframes, self.nframes))
+        if self._epsilon == 'average':
+            self._epsilon = np.zeros((self.nframes, ), )
+            self._type_epsilon = 'average'
+        else:
+            value_epsilon = self._epsilon
+            self._epsilon = np.full((self.nframes, ), value_epsilon)
+            self._type_epsilon = 'constant'
+
+        self._rot = np.zeros(9)
+
+    def _single_frame(self):
+        traj_index = self._ts.frame
+        i_ref = self._u.trajectory[traj_index].positions - self.atoms.center_of_mass()
```

Comment:
They can't update in the explicit loop you have given here. The frame index is just an integer over which you are iterating. There is no interaction with the Reader object involved. The AnalysisBase class will have the following line in the run method.

```
for ts in self.trajectory:
    self._ts = ts
    self._single_frame()
```

the explicit iteration over the trajectory object is updating the timestep and the linked AtomGroups. To update the coordinates of a AtomGroup you have to interact with the Reader object.

But be wary of doing a loop like this. I never tried a loop like this in MDAnalysis and my guess is it will fail if you don't reset the reader after the inner loop.